regparser/tree/struct.py

```diff
@@ -175,27 +181,53 @@ def node_type(self):
     def tagged_text(self):
         return self._tagged_text
 
-    def __repr__(self):
-        if not hasattr(self, '_repr_value'):
-            self._repr_value = (
-                "FrozenNode(text=%s, children=%s, label=%s, title=%s)"
-                % (repr(self.text), repr(self.children), repr(self.label),
-                   repr(self.title)))
-        return self._repr_value
-
-    def __cmp__(self, other):
-        return (other.__class__ == self.__class__
-                and cmp(repr(self), repr(other)))
+    @property
+    def hash(self):
+        return self._hash
+
+    def _generate_hash(self):
+        """Called during instantiation. Digests all fields"""
+        hasher = hashlib.sha256()
+        hasher.update(self.text.encode('utf-8'))
+        hasher.update(self.tagged_text.encode('utf-8'))
+        hasher.update(self.title.encode('utf-8'))
+        hasher.update(self.label_id.encode('utf-8'))
+        hasher.update(self.node_type)
+        for child in self.children:
+            hasher.update(child.hash)
+        return hasher.hexdigest()
 
     def __hash__(self):
-        return hash(repr(self))
+        """As the hash property is already distinctive, re-use it"""
+        return hash(self.hash)
+
+    def __eq__(self, other):
+        """We define equality as having the same fields except for children.
+        Instead of recursively inspecting them, we compare only their hash
+        (this is a Merkle tree)"""
+        return (other.__class__ == self.__class__
+                and self.hash == other.hash
+                and self.text == other.text
```

Comment:
Added a comment to explain; there's a very small chance of a hash collision -- that two different `FrozenNodes` would hash to the same entity. This limits the damage that might do by also comparing fields.